### PR TITLE
Set status error message 

### PR DESCRIPTION
**Pull request by sarahalsmiller**
### Changes proposed in this PR:
- Add RefNotPermitted error to Resolution Errors
- Properly set status when ReferencePolicy doesn't exist

### How We've tested this PR:
- Manually running the conformance tests defined in this PR https://github.com/kubernetes-sigs/gateway-api/pull/1081

### How I expect reviewers to test this PR:
-Note that the conformance test passes
- Manually running the conformance tests if needed

### Checklist:
- [X] Tests added
- [X] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use imperative present tense (e.g. Add support for...)

### CODE REVIEW

File: internal/k8s/reconciler/route.go
Comment (andrewstucki) [inline internal/k8s/reconciler/route.go:208]: Gah, I totally forgot about the need to do this here -- glad you figured it out. For some context for everyone's sake if you're wondering why the code is like this, basically we can't set the status of the route directly when it has something that's invalid.

"Why" you ask? Because according to the spec, routes only have statuses in relation to the parent refs that they specify (yes, this makes route statuses all the more confusing when something is invalid for a route globally, as is the case here).

The result? We memoize any resolution errors from our validation stage onto the struct that we store and then, once binding to a gateway inevitably fails, we dump the resolution errors onto the parentRef-based status.

File: internal/k8s/reconciler/route.go
Comment (andrewstucki) [inline internal/k8s/reconciler/route.go:378]: nit:

```suggestion
					r.resolutionErrors.Add(service.NewRefNotPermittedError(msg))
```

File: internal/k8s/reconciler/route.go
Comment (andrewstucki) [inline internal/k8s/reconciler/route.go:420]: same as above, no real need for the intermediate assignment and the line isn't that long if you just inline the error construction:

```suggestion
			r.resolutionErrors.Add(service.NewRefNotPermittedError(msg))
```

File: internal/k8s/service/resolver.go
Comment (andrewstucki) [inline internal/k8s/service/resolver.go:33]: Thinking we may want to refactor this `ResolutionErrors` struct a bit.

As of now we basically store a bunch of errors that can happen during our resolution process here:

https://github.com/hashicorp/consul-api-gateway/blob/49e3d2fa58ebdef57122d4946c7b99f507400cfe/internal/k8s/service/resolver.go#L72-L79

When we go to set the final `ResolvedRefs` status in the above `OnBindFailed`, we flatten all of the errors that we've had and wrap the flattened error strings into a single "uber error":

https://github.com/hashicorp/consul-api-gateway/blob/49e3d2fa58ebdef57122d4946c7b99f507400cfe/internal/k8s/service/resolver.go#L121-L133

Right now, as written, the `Flatten()` call will result in this branch getting hit:

https://github.com/hashicorp/consul-api-gateway/blob/49e3d2fa58ebdef57122d4946c7b99f507400cfe/internal/k8s/service/resolver.go#L126

due to the lack of `RefNotPermittedErrorType` being taken into account in the error storage on the struct. Since we determine that a ref isn't permitted before resolving any other references on the route, I'm thinking it might make sense just to store the error directly on the struct and if it's not `nil` return the error with a `return RefNotPermittedErrorType, r.refError` in the call to `Flatten()`. Another potential way of handling this is by storing the error for the time being directly on the `K8sRoute` object rather than introducing a new error type here.

If anyone else has other suggestions too I'd be fine with that.

**Review by andrewstucki — COMMENTED** [review]
know this is still a WIP, but leaving some initial feedback that might help with implementation.

File: internal/k8s/reconciler/route.go
Comment (andrewstucki) [reply internal/k8s/reconciler/route.go:208]: Another thing to note, the reason we can't just store statuses on all parent refs for the route prior to attempting the insertion is because during the reference resolution process we don't actually know which gateways that the route references are managed by our controller -- we're not allowed to touch statuses for gateways that aren't managed by us.

File: internal/k8s/service/resolver.go
Comment (sarahalsmiller) [reply internal/k8s/service/resolver.go:33]: I'm cool with implementing that, as long as there isn't going to be a potential situation where we might end up with more than one of them.

Comment (andrewstucki) [reply internal/k8s/service/resolver.go:33]: Actually, that's a good point, since you can have multiple `backendRefs` with multiple required `ReferencePolicy` values, it does make sense to keep track of all of them and give feedback that a user could potentially be missing a few. So, I'm thinking that we can just extend `ResolutionErrors`, the storage arrays, `Add`, and the `Flatten` method. That's probably the most "correct" simplest way of doing things for now.

Thoughts @nathancoleman @mikemorris ?

Comment (nathancoleman) [reply internal/k8s/service/resolver.go:34]: Agree

**Comment by mikemorris** [comment]
Yep, I can work with @sarahalsmiller on writing a changelog entry using the new tooling today!

File: .github/workflows/conformance.yml
Comment (nathancoleman) [inline .github/workflows/conformance.yml:45]:
```suggestion
          ref: "60s-time-to-consistency"
```

I pushed a branch to our hashicorp fork that's name better indicates what it is

File: internal/k8s/reconciler/route_test.go
Comment (andrewstucki) [inline internal/k8s/reconciler/route_test.go:279]: So, I was thinking something more along the lines of this:

```suggestion
	// check route ref
	route = NewK8sRoute(inner, K8sRouteConfig{
		ControllerName: "expected",
		Logger:         hclog.NewNullLogger(),
	})
	route.resolutionErrors.Add(service.NewRefNotPermittedError("not found"))
	route.OnBindFailed(nil, gateway)
	statuses = route.MergedStatus().Parents
	require.Len(t, statuses, 3)
	require.Equal(t, "expected", string(statuses[0].ParentRef.Name))
	require.Equal(t, "expected", string(statuses[0].ControllerName))
	require.Equal(t, "not found", statuses[0].Conditions[1].Message)
	require.Equal(t, RouteConditionReasonRefNotPermitted, statuses[0].Conditions[1].Reason)

	// check binding for non-existent route
	route = NewK8sRoute(inner, K8sRouteConfig{
		ControllerName: "expected",
		Logger:         hclog.NewNullLogger(),
	})
	gateway = NewK8sGateway(&gw.Gateway{
```

Also, looks like you'll need to add some additional error concatenation code here:
https://github.com/hashicorp/consul-api-gateway/blob/a4fc8cd904df96e044563e57946783c818ee4424/internal/k8s/service/resolver.go#L120

since right now there's no casing for `refNotPermittedErrors`

**Review by andrewstucki — CHANGES_REQUESTED** [review]

File: internal/k8s/reconciler/route_test.go
Comment (andrewstucki) [reply internal/k8s/reconciler/route_test.go:279]: just to clarify, the above code didn't actually exercise the case/switch statement added to `OnBindFailed` since there wasn't anything in `resolutionErrors`

File: internal/k8s/reconciler/config/statuses.yaml
Comment (nathancoleman) [inline internal/k8s/reconciler/config/statuses.yaml]: Do we actually use this to bubble up the error to the listener, or is this something we can rip out with the changes in this PR?

File: internal/k8s/reconciler/route.go
Comment (nathancoleman) [inline internal/k8s/reconciler/route.go:377]: I'm used to logs getting piped into systems like Splunk where it's best to put the unique pieces of information in the metadata like this:
```go
r.logger.Warn("Cross-namespace routing not allowed without matching ReferencePolicy", "refName", ref.Name, "refNamespace", ref.Namespace)
```

I'm not sure if this is a pattern that HashiCorp uses but am curious if someone knows

File: internal/k8s/reconciler/route_test.go
Comment (nathancoleman) [inline internal/k8s/reconciler/route_test.go:278]: It's helpful to use the `assert` package whenever failing the assertion doesn't invalidate the rest of the test. In cases like this, it's helpful to know if the later assertions pass/fail even if this one fails, and `require` immediately ends the test on a failure.

File: internal/k8s/reconciler/route_test.go
Comment (nathancoleman) [inline internal/k8s/reconciler/route_test.go:548]: Is it reasonable the assert these things? Comment can be removed either way

**Review by nathancoleman — APPROVED** [review]
Some general preferences but nothing to block merge. Nice work!

File: internal/k8s/reconciler/route.go
Comment (andrewstucki) [reply internal/k8s/reconciler/route.go:377]: Yeah, so the `go-hclog` project we use everywhere (including for our loggers in this project) is a more or less lightweight structured logger, so we should definitely leverage that rather than constructing plain ole strings. Also, if someone turns on [`logJSON`](https://www.consul.io/docs/k8s/helm#v-global-logjson) in `helm` -- using the structured logs will move these into their own fields that can be easily filtered in a centralized log store as you mention.

**Review by andrewstucki — CHANGES_REQUESTED** [review]
Sorry, just one more change for message clarity and to avoid a case where you'd get a panic.

File: internal/k8s/reconciler/route.go
Comment (sarahalsmiller) [reply internal/k8s/reconciler/route.go:376]: It actually shouldn't panic as is (unless ref became nil), because %v just should just print nil, but I'll go ahead and make a quick helper function and update the error message accordingly.

File: internal/k8s/reconciler/route.go
Comment (andrewstucki) [inline internal/k8s/reconciler/route.go:376]: this should say:

```suggestion
					msg := fmt.Sprintf("Cross-namespace routing not allowed without matching ReferencePolicy for Service %q", getServiceID(ref, route.GetNamespace()))
```

as below.

**Review by andrewstucki — APPROVED** [review]
Last fix, but looks good once that's done, so going to approve -- thanks @sarahalsmiller

File: internal/k8s/reconciler/config/statuses.yaml
Comment (mikemorris) [reply internal/k8s/reconciler/config/statuses.yaml]: [For future reference] this has been raised upstream at https://github.com/kubernetes-sigs/gateway-api/issues/1112#issuecomment-1099655762

**Pull request merged**
